### PR TITLE
Replace `in_macro` usage with `from_expansion`

### PR DESCRIPTION
**Pull request by camsteffen**
changelog: none

Generally replace `in_macro(span)` with `span.from_expansion()`. If we're just trying to avoid expanded code, this seems more appropriate because any kind of expanded code is prone to false positives. One place I did not touch is `macro_use.rs`. I think this lint could use a rewrite so I moved `in_macro` there, the only place it is still used.

### CODE REVIEW

**Comment by camsteffen** [comment]
Eh let's wait for the sync

**Comment by flip1995** [comment]
Ah nice, so the for-loop lowering PR got merged and this will get through without test regressions? Nice! I'll make sure to include this PR in the sync-back to rustc 👍

**Review by flip1995 — APPROVED** [review]

**Comment by camsteffen** [comment]
To be clear there are a few changes beyond simply replacing the function, but it's all fairly straightforward.

File: clippy_lints/src/macro_use.rs [hunk 1]
```diff
@@ -213,3 +213,7 @@ impl<'tcx> LateLintPass<'tcx> for MacroUseImports {
         }
     }
 }
+
+fn in_macro(span: Span) -> bool {
+    span.from_expansion() && !matches!(span.ctxt().outer_expn_data().kind, ExpnKind::Desugaring(..))
```
Comment (flip1995) [inline clippy_lints/src/macro_use.rs:218]: Ah, so this is the only place where we can't get rid of the distinction between macro and desugaring, yet.

**Review by flip1995 — APPROVED** [review]
Now also reviewed the other small changes.

**Comment by xFrednet** [comment]
Could you also update the [*Dealing with macros*](https://github.com/rust-lang/rust-clippy/blob/master/doc/common_tools_writing_lints.md#dealing-with-macros) documentation in accordance to this change? It would be good to mention that it's generally better to use `.from_expansion()` :upside_down_face:

**Comment by flip1995** [comment]
> I'll make sure to include this PR in the sync-back to rustc

Since this is a NFC, I decided to rather get the sync through as fast as possible, instead of trying to get this included in the sync.

**Comment by camsteffen** [comment]
Rebased. Removed some redundant `in_external_macro` calls. Edited the docs.

File: clippy_lints/src/macro_use.rs (hunk 1 above)
Comment (camsteffen) [reply clippy_lints/src/macro_use.rs:218]: Right. Actually this lint probably should look for `ExpnKind::Macro(MacroKind::Bang`

File: doc/common_tools_writing_lints.md [hunk 2]
```diff
@@ -182,64 +182,76 @@ impl<'tcx> LateLintPass<'tcx> for MyTypeImpl {
...
+   extern crate a_crate_with_macros;
+
+   // `foo` is defined in `a_crate_with_macros`
+   foo!("bar");
+
+   // if we lint the `match` of `foo` call and test its span
+   assert_eq!(in_external_macro(cx.sess(), match_span), true);
+   ```
+
+   ```rust
```
Comment (flip1995) [inline doc/common_tools_writing_lints.md:238]: That code block seems to be misplaced. Did you mean to remove it? If not, this probably belongs in the `span.ctxt()` section.

Comment (camsteffen) [reply doc/common_tools_writing_lints.md:238]: Oh...I just realized I conflated this PR with another idea I had to remove `differing_macro_contexts`...so what do you think of that? 😄 

For now I'll just restore the lines deleted here.

Comment (camsteffen) [reply doc/common_tools_writing_lints.md:238]: > For now I'll just restore the lines deleted here.

This only makes sense if you view without whitespace. I'll revert changes to the `differing_macro_contexts` section.

**Comment by flip1995** [comment]
Thanks! This looks good now.

About removing `differing_macro_contexts`: IIRC, the only thing that it does is returning `span1.ctxt() != span2.ctxt()`, right? So yeah, having a wrapper function for this doesn't really help complexity-wise. The question is if the implicit documentation of having this function helps. I personally don't think that it is that important, since `span1.ctxt() != span2.ctxt()` is already telling enough IMO. So I guess, it could be removed. But let's do this in an additional PR.

@bors r+ p=1 (codebase wide change)

**Pull request merged**